### ISSUE TITLE
0.21.1: pep517 build packages docs/ content

### ISSUE THREAD

**Issue by kloczek** [python-validators/validators#271]
wheel archive generetaed by:
```console
+ /usr/bin/python3 -sBm build -w --no-isolation
* Getting build dependencies for wheel...
* Building wheel...
Successfully built validators-0.21.1-py3-none-any.whl
```
And it has content:
```console
Archive:  /home/tkloczko/rpmbuild/BUILD/validators-0.21.1/dist/validators-0.21.1-py3-none-any.whl
 Length   Method    Size  Cmpr    Date    Time   CRC-32   Name
--------  ------  ------- ---- ---------- ----- --------  ----
    9778  Defl:N     2891  70% 01-01-1980 00:00 f90038ec  CHANGES.md
     634  Defl:N      348  45% 01-01-1980 00:00 e31fc679  docs/Makefile
    1697  Defl:N      592  65% 01-01-1980 00:00 d7e0f001  docs/conf.py
    5467  Defl:N     1765  68% 01-01-1980 00:00 d7ca5c83  docs/gen_docs.py
    1481  Defl:N      543  63% 01-01-1980 00:00 a780f003  docs/index.md
     800  Defl:N      425  47% 01-01-1980 00:00 fd2f318a  docs/make.bat
    1404  Defl:N      574  59% 01-01-1980 00:00 dfde3183  validators/__init__.py
    1124  Defl:N      372  67% 01-01-1980 00:00 a79a9866  validators/_extremes.py
    3250  Defl:N     1075  67% 01-01-1980 00:00 9d2603cf  validators/between.py
    1739  Defl:N      861  51% 01-01-1980 00:00 e5be3933  validators/btc_address.py
    6270  Defl:N     1047  83% 01-01-1980 00:00 2e8d6f7c  validators/card.py
    1814  Defl:N      789  57% 01-01-1980 00:00 154c3d33  validators/domain.py
    2816  Defl:N     1054  63% 01-01-1980 00:00 3999e63a  validators/email.py
    3606  Defl:N      799  78% 01-01-1980 00:00 a8864406  validators/hashes.py
    3706  Defl:N     1167  69% 01-01-1980 00:00 6c636cd5  validators/hostname.py
     234  Defl:N      145  38% 01-01-1980 00:00 6f9da217  validators/i18n/__init__.py
    5573  Defl:N     1652  70% 01-01-1980 00:00 9bec60a9  validators/i18n/es.py
    3230  Defl:N     1228  62% 01-01-1980 00:00 378215f3  validators/i18n/fi.py
    1146  Defl:N      575  50% 01-01-1980 00:00 16b63bdb  validators/iban.py
    3473  Defl:N      927  73% 01-01-1980 00:00 d8c72d36  validators/ip_address.py
    1175  Defl:N      472  60% 01-01-1980 00:00 4009b667  validators/length.py
     948  Defl:N      476  50% 01-01-1980 00:00 5710b0ff  validators/mac_address.py
       0  Defl:N        2   0% 01-01-1980 00:00 00000000  validators/py.typed
     845  Defl:N      435  49% 01-01-1980 00:00 4ed6c11d  validators/slug.py
    6549  Defl:N     2184  67% 01-01-1980 00:00 112e6b3e  validators/url.py
    2392  Defl:N      896  63% 01-01-1980 00:00 7ca957f1  validators/utils.py
    1147  Defl:N      553  52% 01-01-1980 00:00 fdd3c791  validators/uuid.py
    1089  Defl:N      658  40% 01-01-1980 00:00 74fe0f1f  validators-0.21.1.dist-info/LICENSE
    2799  Defl:N      911  68% 01-01-1980 00:00 e1787640  validators-0.21.1.dist-info/METADATA
      88  Defl:N       86   2% 01-01-1980 00:00 5f07d630  validators-0.21.1.dist-info/WHEEL
    2355  Defl:N     1405  40% 01-01-2016 00:00 5fd0d275  validators-0.21.1.dist-info/RECORD
--------          -------  ---                            -------
   78629            26907  66%                            31 files
```

Looks like MANIFEST.in files needs to be aded

**Comment by joe733** [comment]
Hi @kloczek, what do you mean by `pep517 build packages docs/ content`?

If you build package using `python build_pkg.py` these are the contents of `validators-0.21.1-py3-none-any.whl`:

```console
$ [validators-0.21.1-py3-none-any] ls -laR 
Permissions Size User  Date Modified Name
drwxr-xr-x     - us-er 26 Apr 08:16  docs
drwxr-xr-x     - us-er 26 Apr 08:16  validators
drwxr-xr-x     - us-er 26 Apr 08:16  validators-0.21.1.dist-info
.rw-r--r--  9.8k us-er  1 Jan  1980  CHANGES.md

./docs:
Permissions Size User  Date Modified Name
.rw-r--r--  1.6k us-er  1 Jan  1980  conf.py
.rw-r--r--  5.5k us-er  1 Jan  1980  gen_docs.py
.rw-r--r--  1.5k us-er  1 Jan  1980  index.md
.rwxr-xr-x  1.5k us-er  1 Jan  1980  index.rst
.rw-r--r--   800 us-er  1 Jan  1980  make.bat
.rw-r--r--   634 us-er  1 Jan  1980  Makefile
.rwxr-xr-x   36k us-er  1 Jan  1980  validators.1

./validators:
Permissions Size User  Date Modified Name
drwxr-xr-x     - us-er 26 Apr 08:16  i18n
.rw-r--r--  1.4k us-er  1 Jan  1980  __init__.py
.rw-r--r--  1.1k us-er  1 Jan  1980  _extremes.py
.rw-r--r--  3.2k us-er  1 Jan  1980  between.py
.rw-r--r--  1.7k us-er  1 Jan  1980  btc_address.py
.rw-r--r--  6.3k us-er  1 Jan  1980  card.py
.rw-r--r--  1.8k us-er  1 Jan  1980  domain.py
.rw-r--r--  2.8k us-er  1 Jan  1980  email.py
.rw-r--r--  3.6k us-er  1 Jan  1980  hashes.py
.rw-r--r--  3.7k us-er  1 Jan  1980  hostname.py
.rw-r--r--  1.1k us-er  1 Jan  1980  iban.py
.rw-r--r--  3.5k us-er  1 Jan  1980  ip_address.py
.rw-r--r--  1.2k us-er  1 Jan  1980  length.py
.rw-r--r--   948 us-er  1 Jan  1980  mac_address.py
.rw-r--r--     0 us-er  1 Jan  1980  py.typed
.rw-r--r--   845 us-er  1 Jan  1980  slug.py
.rw-r--r--  6.5k us-er  1 Jan  1980  url.py
.rw-r--r--  2.4k us-er  1 Jan  1980  utils.py
.rw-r--r--  1.1k us-er  1 Jan  1980  uuid.py

./validators/i18n:
Permissions Size User  Date Modified Name
.rw-r--r--   234 us-er  1 Jan  1980  __init__.py
.rw-r--r--  5.6k us-er  1 Jan  1980  es.py
.rw-r--r--  3.2k us-er  1 Jan  1980  fi.py

./validators-0.21.1.dist-info:
Permissions Size User  Date Modified Name
.rw-r--r--  1.1k us-er  1 Jan  1980  LICENSE
.rw-r--r--  2.8k us-er  1 Jan  1980  METADATA
.rw-r--r--  2.5k us-er  1 Jan  2016  RECORD
.rw-r--r--    88 us-er  1 Jan  1980  WHEEL
```

**Comment by joe733** [comment]
These were the content of MANIFEST.in:

```manifest
include CHANGES.rst LICENSE README.rst
recursive-include tests *
recursive-exclude tests *.pyc
recursive-include docs *
recursive-exclude docs *.pyc
prune docs/_build
exclude docs/_themes/.git
```

Commit 6bd8751797355f59a295acbe34d0761986d49509, overrides it with the configuration in `pyproject.toml`:

https://github.com/python-validators/validators/blob/506d22bdecd8a049d701a1ef9efcaf9e9f2ba9aa/pyproject.toml#L43

**Comment by kloczek** [comment]
> Hi @kloczek, what do you mean by `pep517 build packages docs/ content`?

Please try to build .whl archove using `python3 -sBm build -w --no-isolation` command.

> ```toml
> include = ["README.md", "LICENSE", "CHANGES.md", "docs/*", "tests/*"]
> ```

This meand that those files will be included in .whl archive an dinstalled on use `pip` or i`nstaller` module.
Here is the patch against 0.21.1
```patch
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -40,7 +40,7 @@
     "Programming Language :: Python :: Implementation :: CPython",
     "Topic :: Software Development :: Libraries :: Python Modules",
 ]
-include = ["CHANGES.md", "docs/*", "docs/validators.1", "validators/py.typed"]
+include = ["validators/py.typed"]


 ####################
```

**Comment by kloczek** [comment]
> > Please try to build .whl archove using `python3 -sBm build -w --no-isolation` command.
> 
> 1. Why?

Becauseuse pep517 build executed that way will look always the same doesn't matter which one pep517 bakednd will be used.

> 2. I do not see where in [PEP 517](https://peps.python.org/pep-0517), that command is recommended.
>    ```
>    $ python -m build                   
>    /home/user-name/.cache/pypoetry/virtualenvs/validators-Yu3lHuj9-py3.8/bin/python: No module named build
>    ```
> 3. It requires me to install [`build`](https://pypi.org/project/build/).
> 4. You could instead, simply run `python build_pkg.py` which in turn runs `poetry build`.

Yes 'build` is not part of the base putone modules and needs to be installed.

> 5. `validators/py.typed` gets automatically included, why remove others from `.whl`?

Please try to install that .whl using `pip` or `installer` module in some </install/prefix> and than check you will se why.

pep517 is new build procddure.
`build` provudes the same command to execute does't matter which one backend is used (in your case it is `seruptools`).

**Comment by kloczek** [comment]
https://peps.python.org/pep-0517/

**Comment by kloczek** [comment]
> My build backed is `poetry` not `setuptools`.
> 
> https://github.com/python-validators/validators/blob/506d22bdecd8a049d701a1ef9efcaf9e9f2ba9aa/pyproject.toml#L9-L13
> 
> I commented out the following in `pyproject.toml`:
> 
> ```toml
> # include = ["README.md", "LICENSE", "CHANGES.md", "docs/*", "tests/*"]
> ```

There are two possibilies of include/exclude.
One includes/excludes from sdist and whl, and second one only include/exclude from whl.
This one causes that those files are included in whl as well.

> 
> > Please try to install that .whl using pip or installer module in some </install/prefix> and than check you will se why.
> 
> Then I did, but what am I looking for?

Into destdir passed as param for `pip` or `installer`.

> ```
> $ ls -la
> Permissions Size User  Date Modified Name
> drwxr-xr-x     - us-er 26 Apr 14:50  .venv
> .rw-r--r--   23k us-er 26 Apr 14:50  validators-0.21.1-py3-none-any.whl
> 
> $ pip install validators-0.21.1-py3-none-any.whl 
> Processing ./validators-0.21.1-py3-none-any.whl
> Installing collected packages: validators
> Successfully installed validators-0.21.1
> 
> $ pip list                    
> Package    Version
> ---------- -------
> pip        23.1.1
> setuptools 67.7.2
> validators 0.21.1
> 
> $ ls -la .venv/lib/python3.8/site-packages/
> Permissions Size User  Date Modified Name
> drwxr-xr-x     - us-er 26 Apr 14:51  _distutils_hack
> drwxr-xr-x     - us-er 26 Apr 14:51  pip
> drwxr-xr-x     - us-er 26 Apr 14:51  pip-23.1.1.dist-info
> drwxr-xr-x     - us-er 26 Apr 14:51  pkg_resources
> drwxr-xr-x     - us-er 26 Apr 14:51  setuptools
> drwxr-xr-x     - us-er 26 Apr 14:51  setuptools-67.7.2.dist-info
> drwxr-xr-x     - us-er 26 Apr 14:51  validators
> drwxr-xr-x     - us-er 26 Apr 14:51  validators-0.21.1.dist-info
> .rw-r--r--   151 us-er 26 Apr 14:51  distutils-precedence.pth
> 
> $ ls -la .venv/lib/python3.8/site-packages/validators/
> Permissions Size User  Date Modified Name
> drwxr-xr-x     - us-er 26 Apr 14:51  __pycache__
> drwxr-xr-x     - us-er 26 Apr 14:51  i18n
> .rw-r--r--  1.4k us-er 26 Apr 14:51  __init__.py
> .rw-r--r--  1.1k us-er 26 Apr 14:51  _extremes.py
> .rw-r--r--  3.2k us-er 26 Apr 14:51  between.py
> .rw-r--r--  1.7k us-er 26 Apr 14:51  btc_address.py
> .rw-r--r--  6.3k us-er 26 Apr 14:51  card.py
> .rw-r--r--  1.8k us-er 26 Apr 14:51  domain.py
> .rw-r--r--  2.8k us-er 26 Apr 14:51  email.py
> .rw-r--r--  3.6k us-er 26 Apr 14:51  hashes.py
> .rw-r--r--  3.7k us-er 26 Apr 14:51  hostname.py
> .rw-r--r--  1.1k us-er 26 Apr 14:51  iban.py
> .rw-r--r--  3.5k us-er 26 Apr 14:51  ip_address.py
> .rw-r--r--  1.2k us-er 26 Apr 14:51  length.py
> .rw-r--r--   948 us-er 26 Apr 14:51  mac_address.py
> .rw-r--r--     0 us-er 26 Apr 14:51  py.typed
> .rw-r--r--   845 us-er 26 Apr 14:51  slug.py
> .rw-r--r--  6.5k us-er 26 Apr 14:51  url.py
> .rw-r--r--  2.4k us-er 26 Apr 14:51  utils.py
> .rw-r--r--  1.1k us-er 26 Apr 14:51  uuid.py
> ```
> 
> > ... does't matter which one backend is used ...

Which one `pip` version you are using?
Can you try to use it without venv and execurte pip with `--dest /tmp/foo` and check what is inside?

> If so, then why not use `poetry build`?

Because in exact case you need to know first that exct module uses for example `poetry` as pep517 backend.
With `build` you can execute exacly the same command for setuptools, poetry, whey, hatch, hatclings, setuptools-scm, pdm and few other already available.

> [PEP 517](https://peps.python.org/pep-0517/) is a specification that decouples frontends and backends in Python packaging. You can refer this talk ![yt_favicon](https://camo.githubusercontent.com/a58eee0c747d0e16f122025609bdc500d60e63b876cadb74ea0de34bb02808e6/68747470733a2f2f7777772e796f75747562652e636f6d2f732f6465736b746f702f65396136376463642f696d672f66617669636f6e2e69636f) [What does PEP 517 mean for packaging](https://www.youtube.com/watch?v=s5lJsFzv_iI). The goal was to "dethrone distutils and setuptools as the single blessed way of making Python packages."
> 
> On another note, you're typing like an email spammer.

What you mean?  😋

**Comment by joe733** [comment]
Are you looking for this?

```toml
include = [
    { path = "CHANGES.md", format = "sdist" },
    { path = "docs/*", format = "sdist" },
    { path = "tests/*", format = "sdist" },
]
```

ref: https://python-poetry.org/docs/pyproject#include-and-exclude

---

**`sdist contents`**

```console
$ ls -laR ./validators-0.21.1/
Permissions Size User  Date Modified Name
drwxr-xr-x     - us-er 27 Apr 08:36  docs
drwxr-xr-x     - us-er 27 Apr 08:36  tests
drwxr-xr-x     - us-er 27 Apr 08:36  validators
.rw-r--r--  9.8k us-er 12 Apr 19:02  CHANGES.md
.rw-r--r--  1.1k us-er 31 Mar 21:39  LICENSE
.rw-r--r--  2.8k us-er  1 Jan  1970  PKG-INFO
.rw-r--r--  3.9k us-er 27 Apr 08:34  pyproject.toml
.rw-r--r--  1.5k us-er 31 Mar 21:39  README.md

./validators-0.21.1/docs:
Permissions Size User  Date Modified Name
.rw-r--r--  1.6k us-er 31 Mar 21:39  conf.py
.rw-r--r--  5.5k us-er 24 Apr 15:34  gen_docs.py
.rw-r--r--  1.5k us-er 27 Apr 08:36  index.md
.rw-r--r--  1.5k us-er 27 Apr 08:36  index.rst
.rw-r--r--   800 us-er 31 Mar 21:39  make.bat
.rw-r--r--   634 us-er 31 Mar 21:39  Makefile
.rwxr-xr-x   36k us-er 27 Apr 08:36  validators.1

./validators-0.21.1/tests:
Permissions Size User  Date Modified Name
drwxr-xr-x     - us-er 27 Apr 08:36  i18n
.rw-r--r--    57 us-er 31 Mar 21:39  __init__.py
.rw-r--r--  1.4k us-er 31 Mar 21:39  test__extremes.py
.rw-r--r--  1.1k us-er 31 Mar 21:39  test_between.py
.rw-r--r--  1.2k us-er 31 Mar 21:39  test_btc_address.py
.rw-r--r--  4.4k us-er 31 Mar 21:39  test_card.py
.rw-r--r--  1.9k us-er 31 Mar 21:39  test_domain.py
.rw-r--r--  1.5k us-er 31 Mar 21:39  test_email.py
.rw-r--r--  4.3k us-er 31 Mar 21:39  test_hashes.py
.rw-r--r--  2.6k us-er 31 Mar 21:39  test_hostname.py
.rw-r--r--   580 us-er 31 Mar 21:39  test_iban.py
.rw-r--r--  2.6k us-er 31 Mar 21:39  test_ip_address.py
.rw-r--r--   853 us-er 31 Mar 21:39  test_length.py
.rw-r--r--   848 us-er 31 Mar 21:39  test_mac_address.py
.rw-r--r--   708 us-er 31 Mar 21:39  test_slug.py
.rw-r--r--  6.6k us-er 12 Apr 19:02  test_url.py
.rw-r--r--   965 us-er 31 Mar 21:39  test_uuid.py
.rw-r--r--   959 us-er 31 Mar 21:39  test_validation_failure.py

./validators-0.21.1/tests/i18n:
Permissions Size User  Date Modified Name
.rw-r--r--    61 us-er 31 Mar 21:39  __init__.py
.rw-r--r--  2.8k us-er 31 Mar 21:39  test_es.py
.rw-r--r--  2.1k us-er 31 Mar 21:39  test_fi.py

./validators-0.21.1/validators:
Permissions Size User  Date Modified Name
drwxr-xr-x     - us-er 27 Apr 08:36  i18n
.rw-r--r--  1.4k us-er 12 Apr 19:02  __init__.py
.rw-r--r--  1.1k us-er 31 Mar 21:39  _extremes.py
.rw-r--r--  3.2k us-er 31 Mar 21:39  between.py
.rw-r--r--  1.7k us-er 31 Mar 21:39  btc_address.py
.rw-r--r--  6.3k us-er 31 Mar 21:39  card.py
.rw-r--r--  1.8k us-er 31 Mar 21:39  domain.py
.rw-r--r--  2.8k us-er 31 Mar 21:39  email.py
.rw-r--r--  3.6k us-er 31 Mar 21:39  hashes.py
.rw-r--r--  3.7k us-er 31 Mar 21:39  hostname.py
.rw-r--r--  1.1k us-er 31 Mar 21:39  iban.py
.rw-r--r--  3.5k us-er 31 Mar 21:39  ip_address.py
.rw-r--r--  1.2k us-er 31 Mar 21:39  length.py
.rw-r--r--   948 us-er 31 Mar 21:39  mac_address.py
.rw-r--r--     0 us-er 31 Mar 21:39  py.typed
.rw-r--r--   845 us-er 31 Mar 21:39  slug.py
.rw-r--r--  6.5k us-er 12 Apr 19:02  url.py
.rw-r--r--  2.4k us-er 31 Mar 21:39  utils.py
.rw-r--r--  1.1k us-er 31 Mar 21:39  uuid.py

./validators-0.21.1/validators/i18n:
Permissions Size User  Date Modified Name
.rw-r--r--   234 us-er 31 Mar 21:39  __init__.py
.rw-r--r--  5.6k us-er 31 Mar 21:39  es.py
.rw-r--r--  3.2k us-er 31 Mar 21:39  fi.py
```

---

**`whl contents`**

```console
$ ls -laR ./validators-0.21.1-py3-none-any/
Permissions Size User  Date Modified Name
drwxr-xr-x     - us-er 27 Apr 08:39  validators
drwxr-xr-x     - us-er 27 Apr 08:39  validators-0.21.1.dist-info

./validators-0.21.1-py3-none-any/validators:
Permissions Size User  Date Modified Name
drwxr-xr-x     - us-er 27 Apr 08:39  i18n
.rw-r--r--  1.4k us-er  1 Jan  1980  __init__.py
.rw-r--r--  1.1k us-er  1 Jan  1980  _extremes.py
.rw-r--r--  3.2k us-er  1 Jan  1980  between.py
.rw-r--r--  1.7k us-er  1 Jan  1980  btc_address.py
.rw-r--r--  6.3k us-er  1 Jan  1980  card.py
.rw-r--r--  1.8k us-er  1 Jan  1980  domain.py
.rw-r--r--  2.8k us-er  1 Jan  1980  email.py
.rw-r--r--  3.6k us-er  1 Jan  1980  hashes.py
.rw-r--r--  3.7k us-er  1 Jan  1980  hostname.py
.rw-r--r--  1.1k us-er  1 Jan  1980  iban.py
.rw-r--r--  3.5k us-er  1 Jan  1980  ip_address.py
.rw-r--r--  1.2k us-er  1 Jan  1980  length.py
.rw-r--r--   948 us-er  1 Jan  1980  mac_address.py
.rw-r--r--     0 us-er  1 Jan  1980  py.typed
.rw-r--r--   845 us-er  1 Jan  1980  slug.py
.rw-r--r--  6.5k us-er  1 Jan  1980  url.py
.rw-r--r--  2.4k us-er  1 Jan  1980  utils.py
.rw-r--r--  1.1k us-er  1 Jan  1980  uuid.py

./validators-0.21.1-py3-none-any/validators/i18n:
Permissions Size User  Date Modified Name
.rw-r--r--   234 us-er  1 Jan  1980  __init__.py
.rw-r--r--  5.6k us-er  1 Jan  1980  es.py
.rw-r--r--  3.2k us-er  1 Jan  1980  fi.py

./validators-0.21.1-py3-none-any/validators-0.21.1.dist-info:
Permissions Size User  Date Modified Name
.rw-r--r--  1.1k us-er  1 Jan  1980  LICENSE
.rw-r--r--  2.8k us-er  1 Jan  1980  METADATA
.rw-r--r--  1.9k us-er  1 Jan  2016  RECORD
.rw-r--r--    88 us-er  1 Jan  1980  WHEEL
```

**Comment by joe733** [comment]
Fixed in #283
